typing/typecore.ml

```diff
@@ -5165,13 +5170,21 @@ and type_label_exp create env rmode loc ty_expected
   let arg =
     let snap = if vars = [] then None else Some (Btype.snapshot ()) in
     let arg_mode =
-      match label.lbl_global with
-      | Global -> Value_mode.global
-      | Nonlocal -> Value_mode.local_to_regional rmode
-      | Unrestricted -> rmode
+      match label.lbl_global, label.lbl_repres with
+      | Global, _ ->
+         mode_global
+      | Nonlocal, _ ->
+         expected_mode.mode
+         |> Value_mode.regional_to_global
+         |> Value_mode.local_to_regional
+         |> mode_nontail
```

Comment:
Would you mind wrapping this up as something like `mode_nonlocal_subcomponent` and put it with all the other `mode_foo` functions.

Reply:
Actually, I think that's probably unsafe. It would allow something like:
```ocaml
type 'a t = { nonlocal_ escape : 'a } [@@unboxed]

let local_to_nonlocal : local_ 'a -> local_ 'a t =
  fun x -> { escape = x }

let leak x =
  let r = local_ ref x in
  let q = local_to_nonlocal r in
  q.escape
```

Reply:
Fixed, the call to `Value_mode.regional_to_global` is back in now.